[RPC Framework] Supporting reading the input from the remote worker

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56943 [RPC Framework] Supporting reading the input from the remote worker**
* #56853 Clang format dist_utils.py and rpc/__init__.py

If the module is placed on a CUDA device, then all the CPU tensors in `args` and `kwargs` will also be implicitly moved to the same CUDA device to run forward.

Currently still need to move the forward output from CUDA device back to CPU, until:
1) Process group RPC backend is completely deprecated, and we always use TensorPipe RPC backend;
2) A device map is explicitly provided to TensorPipe RPC backend.

These steps will be done in a separate PR.

#Original PR issue: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D27934791](https://our.internmc.facebook.com/intern/diff/D27934791/)

Commit:
[RPC Framework] Supporting reading the input from the remote worker

If the module is placed on a CUDA device, then all the CPU tensors in `args` and `kwargs` will also be implicitly moved to the same CUDA device to run forward.

Currently still need to move the forward output from CUDA device back to CPU, until:
1) Process group RPC backend is completely deprecated, and we always use TensorPipe RPC backend;
2) A device map is explicitly provided to TensorPipe RPC backend.

These steps will be done in a separate PR.

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D27934791](https://our.internmc.facebook.com/intern/diff/D27934791/)

[ghstack-poisoned]

wanchaol commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -479,3 +479,33 @@ def test_invalid_devices(self):
+        t2 = t1 * 2
+        kwargs = dict(word=t2)
+
+        # Only test Python nn.Module, because script module methods don't support taking kwargs.
```
can we test the script mode without specifying kwargs?

SciPioneer replied:
Added `test_input_moved_to_cuda_device_script`

wanchaol reviewed (approved):
looks good to me, thanks!

rohan-varma commented on torch/distributed/nn/jit/templates/remote_module_template.py:
```diff
@@ -13,13 +13,47 @@
+    args = ({args},)
+    out_args = ()
+    for arg in args:
+        arg = (arg.to(device),) if isinstance(arg, Tensor) else (arg,)
```
What if `args` contains data structures with nested tensors, such as `List[tensor]` or `Dict[str, tensor]`? IIUC torchscript supports those types so we may want to extract tensors from them and move those too?  Although this is an enhancement and can probably be saved for a follow up PR

SciPioneer replied:
Good point! A warning is added here.

I am not sure if there is any real use case that passes nested tensors in args or kwargs. Have you seen such an example in a real use case? If not, I prefer not overcomplicate the implementation at this time.

rohan-varma replied:
I have encountered a few use cases (e.g. internal hpc workload) where the return type contains `List[Tensor]` so users may hit this when training more complex model. But agreed, I don't think it's a hard blocker on this PR.

Worth mentioning in passing - it is actually quite easy to unpack/repack tensors from nested data structure, see pytree utils: https://github.com/pytorch/pytorch/blob/master/torch/utils/_pytree.py

rohan-varma commented on torch/distributed/nn/jit/templates/remote_module_template.py:
```diff
@@ -13,13 +13,47 @@
+    kwargs = {{{kwargs}}}
+    for k, v in kwargs.items():
+        if isinstance(v, Tensor):
+            kwargs[k] = kwargs[k].to(device)
```
Same comment above re: nested args probably applies to kwargs as well.

rohan-varma commented on torch/distributed/nn/jit/templates/remote_module_template.py:
```diff
@@ -13,13 +13,47 @@
+    # do no move any input on cuda devices back to CPU.
+    ret = ()
+    for arg in module.forward(*out_args, {kwargs}):
+        arg = (arg.cpu(),) if isinstance(arg, Tensor) else (arg,)
```
Same comment above re: nested args probably applies to nested module forward() outputs as well.

rohan-varma commented on torch/distributed/nn/jit/templates/remote_module_template.py:
```diff
@@ -13,13 +13,47 @@
+            kwargs[k] = kwargs[k].to(device)
+
+    # Since only CPU tensors are allowed to send over wire,
+    # need to move any GPU tensor to CPU in the output.
```
Do we get an error currently if we leave the outputs on GPU when the backend is tensorpipe? IIUC, it should be okay to leave the outputs on GPU, and TP agent will ensure that on the caller they will be on the appropriate cuda device for that worker via the global device_maps. Is there any reason we can't use this functionality here?

SciPioneer replied:
To leave the output on GPU, TensorPipe backend requires passing an explicit device map.

Otherwise, it will throw an error. See:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L1460

rohan-varma replied:
I see, that's interesting. @mrshenli do you know why the global device maps aren't picked up here, and is that an expected occurence?

SciPioneer commented on torch/distributed/nn/jit/templates/remote_module_template.py:
```diff
@@ -13,13 +13,47 @@
+        arg = (arg.to(device),) if isinstance(arg, Tensor) else (arg,)
+        out_args = out_args + arg
+
+    kwargs = {{{kwargs}}}
```
I followed the example in `forward_async` and `forward_sync`.

My guess is that we need to create a dictionary here, and the inner 2 layers of brackets can be viewed as a unpacking operator.

Commit:
Update on "[RPC Framework] Supporting reading the input from the remote worker"


If the module is placed on a CUDA device, then all the CPU tensors in `args` and `kwargs` will also be implicitly moved to the same CUDA device to run forward.

Currently still need to move the forward output from CUDA device back to CPU, until:
1) Process group RPC backend is completely deprecated, and we always use TensorPipe RPC backend;
2) A device map is explicitly provided to TensorPipe RPC backend.

These steps will be done in a separate PR.

#Original PR issue: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D27934791](https://our.internmc.facebook.com/intern/diff/D27934791/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Supporting reading the input from the remote worker"


If the module is placed on a CUDA device, then all the CPU tensors in `args` and `kwargs` will also be implicitly moved to the same CUDA device to run forward.

Currently still need to move the forward output from CUDA device back to CPU, until:
1) Process group RPC backend is completely deprecated, and we always use TensorPipe RPC backend;
2) A device map is explicitly provided to TensorPipe RPC backend.

These steps will be done in a separate PR.

#Original PR issue: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D27934791](https://our.internmc.facebook.com/intern/diff/D27934791/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Supporting reading the input from the remote worker"


If the module is placed on a CUDA device, then all the CPU tensors in `args` and `kwargs` will also be implicitly moved to the same CUDA device to run forward.

Currently still need to move the forward output from CUDA device back to CPU, until:
1) Process group RPC backend is completely deprecated, and we always use TensorPipe RPC backend;
2) A device map is explicitly provided to TensorPipe RPC backend.

These steps will be done in a separate PR.

#Original PR issue: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D27934791](https://our.internmc.facebook.com/intern/diff/D27934791/)

[ghstack-poisoned]
